connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java

```diff
@@ -0,0 +1,227 @@
+    }
+
+    /**
+     * Stop and unregister the metrics from any reporters.
+     */
+    public void stop() {
+        metrics.close();
+        LOG.debug("Unregistering Connect metrics with JMX for worker '{}'", workerId);
+        AppInfoParser.unregisterAppInfo(JMX_PREFIX, workerId);
+    }
+
+    /**
+     * A group of metrics. Each group maps to a JMX MBean and each metric maps to an MBean attribute.
+     */
+    public class MetricGroup {
+        private final String groupName;
+        private final Map<String, String> tags;
+
+        protected MetricGroup(String groupName, Map<String, String> tags) {
+            this.groupName = groupName;
+            this.tags = tags;
+        }
+
+        /**
+         * Create the name of a metric that belongs to this group and has the group's tags.
+         *
+         * @param name the name of the metric/attribute; may not be null
+         * @param desc the description for the metric/attribute; may not be null
+         * @return the metric name; never null
+         */
+        public MetricName metricName(String name, String desc) {
+            return metrics.metricName(name, groupName, desc, tags);
+        }
+
+        /**
+         * The {@link Metrics} that this group belongs to.
+         *
+         * @return the metrics; never null
+         */
+        public Metrics metrics() {
+            return metrics;
+        }
+
+        Map<String, String> tags() {
+            return Collections.unmodifiableMap(tags);
+        }
+    }
+
+    /**
+     * Create a set of tags using the supplied key and value pairs.
+     *
+     * @param workerId the worker ID that should be included first in the tags; may be null if not to be included
+     * @param keyValue the key and value pairs for the tags; must be an even number
+     * @return the map of tags that can be supplied to the {@link Metrics} methods; never null
+     */
+    static Map<String, String> tags(String workerId, String... keyValue) {
+        if ((keyValue.length % 2) != 0)
+            throw new IllegalArgumentException("keyValue needs to be specified in pairs");
+        Map<String, String> tags = new HashMap<>();
+        if (workerId != null && !workerId.trim().isEmpty()) {
+            tags.put(WORKER_ID_TAG_NAME, workerId);
+        }
+        for (int i = 0; i < keyValue.length; i += 2) {
+            tags.put(keyValue[i], keyValue[i + 1]);
+        }
+        return tags;
+    }
+
+    /**
+     * Utility to ensure the supplied name contains valid characters, replacing sequences of 1 or more
+     * ':', '/', '\', '*', '?', '[', ']', '=', or comma characters with a single '-'.
+     *
+     * @param name the name; may not be null
+     * @return the validated name; never null
+     */
+    static String makeValidName(String name) {
+        Objects.requireNonNull(name);
+        name = name.trim();
+        if (!name.isEmpty()) {
+            name = name.replaceAll("[:/\\\\*?,;\\[\\]\\\\=]+", "-");
```

Comment:
I know the naming thing has bit us in the past, is this same approach used elsewhere and/or how was it decided on? Specifically, metric name constraints really shouldn't be JMX specific if that is the case here, despite the fact that the metrics is so obviously JMX-inspired.

I can easily find https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/metrics/KafkaMetricsGroup.scala#L46 but nothing else. Have we not had the same problems because metrics w/ topic names in them already  have constraints on the naming?

If I am remembering correctly, I think maybe both @gwenshap and @junrao were involved in some discussions, I think `-` vs `_` was a problem at some point? Maybe one of them could chime in here.

Reply:
Another way of asking this is, should this actually happen in Connect or should it be happening in `JmxReporter`?

Reply:
I guess a question I have is whether the metric names must be validated to satisfy all reporters. If so, then since the metrics are created before the reporters are, don't we need to ensure the metrics satisfy all reporters' requirements. That seems onerous.

Another option that may or may not be practical is to expect each reporter to accept metric names with whatever limitations we allow, and to deal with any conversions if that isn't the case.

I'm happy to move this into a more reusable place and/or to use existing conventions/rules.

Reply:
to me it seems like we can't possibly know what the constraints of all reporters would be and they don't provide an interface for validation, so it should be up to them to figure out how to substitute. but i've also asked some other folks to maybe chime in here who may have better context on how we've handled this elsewhere.

Reply:
We did the dot to _ conversion for Yammer metric mostly because reporters like Graphite typically use dot to represent hierarchy and quite a few people are using the existing Graphite reporter that may be confused with dot.

Since Kafka metric is new, we could just let individual reporter deal with this issue, instead of changing the metric name directly.

Reply:
Oh, I did forget this: the `workerId` is current set to:

    advertisedUrl.getHost() + ":" + advertisedUrl.getPort()

which means that at least `:` will be in one of the names that make up the group name, and JMX names cannot have `:` characters. So right off the bat, we're using names that aren't valid in JMX.

Reply:
Actually, I take that back. The workerId will be used in the tags for a few of the Connect metric groups. The only restriction we have on connector names is in distributed mode (may not contain `/` character). So it'd be pretty easy to break JMX naming conventions.